File: generate_killer.cpp
/*
 * generate_killer.cpp
 * High-performance Killer Sudoku Generator
 *
 * Compile with: g++ -O3 -o generate_killer generate_killer.cpp
 * Run with: ./generate_killer --count 5
 */

#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <algorithm>
#include <numeric>
#include <set>
#include <map>
#include <chrono>
#include <fstream>
#include <cstring>
#include <cassert>
#include <iomanip>

using namespace std;

// ============================================================================
// Constants & Globals
// ============================================================================

const int N = 9;
const int TOTAL_CELLS = 81;
const char* OUTPUT_FILE = "sudoku_killer.json";

struct Cage {
    vector<int> cells; // 0-80 indices
    int sum;
};

struct Difficulty {
    string name;
    int min_cage;
    int max_cage;
    double target_avg;
};

map<int, Difficulty> DIFFICULTY_LEVELS = {
    {1, {"Beginner", 2, 4, 2.5}},
    {2, {"Easy",     2, 4, 2.8}},
    {3, {"Medium",   2, 4, 3.0}},
    {4, {"Hard",     2, 4, 3.2}},
    {5, {"Expert",   2, 4, 3.5}}
};

// Random engine
mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());

// ============================================================================
// Helper Functions
// ============================================================================

int get_random_int(int min, int max) {
    uniform_int_distribution<int> dist(min, max);
    return dist(rng);
}

// Union-Find Data Structure
struct UnionFind {
    vector<int> parent;
    vector<int> size;

    UnionFind(int n) {
        parent.resize(n);
        iota(parent.begin(), parent.end(), 0);
        size.assign(n, 1);
    }

    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    bool unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) swap(ra, rb);
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }
};

// ============================================================================
// Fast Solver (Integrated from killer_solver.c)
// ============================================================================

class KillerSolver {
    int cell_allowed[TOTAL_CELLS];
    int cell_cage[TOTAL_CELLS];
    int row_used[N], col_used[N], box_used[N];
    int cage_used[TOTAL_CELLS], cage_partial[TOTAL_CELLS], cage_unfilled[TOTAL_CELLS];
    int cage_sum[TOTAL_CELLS];
    bool assigned[TOTAL_CELLS];
    
    int solutions;
    int num_cages;

    inline int popcnt(int x) {
        return __builtin_popcount(x);
    }

    inline int ctz(int x) {
        return __builtin_ctz(x);
    }

    void solve_recursive() {
        if (solutions >= 2) return;

        // Dynamic MRV
        int best = -1, best_cnt = 100;
        for (int i = 0; i < TOTAL_CELLS; i++) {
            if (assigned[i]) continue;
            int r = i / N, c = i % N, bx = (r / 3) * 3 + c / 3;
            int ci = cell_cage[i];
            int avail = cell_allowed[i] & ~(row_used[r] | col_used[c] | box_used[bx] | cage_used[ci]);
            int cnt = popcnt(avail);
            if (cnt == 0) return;
            if (cnt < best_cnt) {
                best_cnt = cnt;
                best = i;
                if (cnt == 1) break;
            }
        }

        if (best < 0) {
            solutions++;
            return;
        }

        int i = best;
        int r = i / N, c = i % N, bx = (r / 3) * 3 + c / 3;
        int ci = cell_cage[i];
        int avail = cell_allowed[i] & ~(row_used[r] | col_used[c] | box_used[bx] | cage_used[ci]);

        assigned[i] = true;
        while (avail) {
            int bit = avail & (-avail);
            avail ^= bit;
            int digit = ctz(bit); // 1..9 based on bit position

            int np = cage_partial[ci] + digit;
            int nu = cage_unfilled[ci] - 1;
            
            // Pruning
            if (np > cage_sum[ci]) continue;
            if (nu == 0) {
                if (np != cage_sum[ci]) continue;
            } else {
                int rem = cage_sum[ci] - np;
                // Min possible sum for 'nu' remaining digits is 1+2+...+nu? 
                // No, distinct digits constraint is handled by mask, but we can do a simple range check
                // Actually, just checking if rem is possible is complex, simple check:
                if (rem < (nu * (nu + 1)) / 2) continue; // Basic min sum check
            }

            row_used[r] |= bit;
            col_used[c] |= bit;
            box_used[bx] |= bit;
            cage_used[ci] |= bit;
            cage_partial[ci] = np;
            cage_unfilled[ci] = nu;

            solve_recursive();

            row_used[r] ^= bit;
            col_used[c] ^= bit;
            box_used[bx] ^= bit;
            cage_used[ci] ^= bit;
            cage_partial[ci] -= digit;
            cage_unfilled[ci]++;

            if (solutions >= 2) break;
        }
        assigned[i] = false;
    }

public:
    bool has_unique_solution(const vector<Cage>& cages) {
        num_cages = cages.size();
        memset(cell_allowed, 0, sizeof(cell_allowed));
        memset(cell_cage, 0, sizeof(cell_cage));
        memset(cage_partial, 0, sizeof(cage_partial));
        memset(cage_used, 0, sizeof(cage_used));

        for (int ci = 0; ci < num_cages; ci++) {
            int sz = cages[ci].cells.size();
            int sm = cages[ci].sum;
            cage_sum[ci] = sm;
            cage_unfilled[ci] = sz;

            int mask_union = 0;
            // Precompute valid combos for this cage
            if (sz == 1) {
                if (sm >= 1 && sm <= 9) mask_union = 1 << sm;
            } else {
                int start = (1 << sz) - 1;
                for (int s = start; s < (1 << 10); ) {
                    if ((s & 1) == 0 && popcnt(s) == sz) {
                        int sum_val = 0;
                        int tmp = s;
                        while (tmp) {
                            int b = tmp & (-tmp);
                            sum_val += ctz(b);
                            tmp ^= b;
                        }
                        if (sum_val == sm) mask_union |= s;
                    }
                    int c2 = s & (-s);
                    int r2 = s + c2;
                    s = (((r2 ^ s) >> 2) / c2) | r2;
                }
            }

            if (mask_union == 0) return false; // Impossible cage

            for (int cell_idx : cages[ci].cells) {
                cell_allowed[cell_idx] = mask_union;
                cell_cage[cell_idx] = ci;
            }
        }

        memset(row_used, 0, sizeof(row_used));
        memset(col_used, 0, sizeof(col_used));
        memset(box_used, 0, sizeof(box_used));
        memset(assigned, 0, sizeof(assigned));
        solutions = 0;

        solve_recursive();
        return solutions == 1;
    }
};

KillerSolver solver;

// ============================================================================
// Board Generator
// ============================================================================

bool is_valid_board(const vector<vector<int>>& board, int r, int c, int num) {
    for (int k = 0; k < 9; k++) if (board[r][k] == num) return false;
    for (int k = 0; k < 9; k++) if (board[k][c] == num) return false;
    int br = (r / 3) * 3, bc = (c / 3) * 3;
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[br + i][bc + j] == num) return false;
    return true;
}

bool fill_board(vector<vector<int>>& board, int pos) {
    if (pos == 81) return true;
    int r = pos / 9, c = pos % 9;
    
    vector<int> nums(9);
    iota(nums.begin(), nums.end(), 1);
    shuffle(nums.begin(), nums.end(), rng);

    for (int n : nums) {
        if (is_valid_board(board, r, c, n)) {
            board[r][c] = n;
            if (fill_board(board, pos + 1)) return true;
            board[r][c] = 0;
        }
    }
    return false;
}

vector<vector<int>> generate_full_board() {
    vector<vector<int>> board(9, vector<int>(9, 0));
    fill_board(board, 0);
    return board;
}

// ============================================================================
// Cage Generator
// ============================================================================

vector<Cage> cages_from_uf(UnionFind& uf, const vector<vector<int>>& board) {
    map<int, vector<int>> groups;
    for (int i = 0; i < TOTAL_CELLS; i++) {
        groups[uf.find(i)].push_back(i);
    }
    vector<Cage> cages;
    for (auto& kv : groups) {
        Cage c;
        c.cells = kv.second;
        c.sum = 0;
        for (int idx : c.cells) {
            c.sum += board[idx / 9][idx % 9];
        }
        cages.push_back(c);
    }
    return cages;
}

vector<Cage> generate_cages(const vector<vector<int>>& board, int difficulty) {
    Difficulty info = DIFFICULTY_LEVELS[difficulty];
    int target_count = max((int)(TOTAL_CELLS / info.target_avg), 9);
    
    // Edges
    vector<pair<int, int>> edges;
    for (int r = 0; r < 9; r++) {
        for (int c = 0; c < 9; c++) {
            int idx = r * 9 + c;
            if (c + 1 < 9) edges.push_back({idx, idx + 1});
            if (r + 1 < 9) edges.push_back({idx, idx + 9});
        }
    }

    UnionFind uf(TOTAL_CELLS);
    
    // Phase 1: Merge iteratively
    for (int round = 0; round < 20; round++) {
        // Shuffle first to ensure randomness among equal-sized merges
        shuffle(edges.begin(), edges.end(), rng);
        
        // Sort edges by combined size of cages they connect to prioritize smaller merges
        stable_sort(edges.begin(), edges.end(), [&](const pair<int,int>& a, const pair<int,int>& b) {
            // Check bounds (defensive)
            if (a.first < 0 || a.first >= TOTAL_CELLS || a.second < 0 || a.second >= TOTAL_CELLS) return false;
            if (b.first < 0 || b.first >= TOTAL_CELLS || b.second < 0 || b.second >= TOTAL_CELLS) return false;
            
            int root_a1 = uf.find(a.first);
            int root_a2 = uf.find(a.second);
            int root_b1 = uf.find(b.first);
            int root_b2 = uf.find(b.second);
            
            int size_a = uf.size[root_a1] + uf.size[root_a2];
            int size_b = uf.size[root_b1] + uf.size[root_b2];
            return size_a < size_b;
        });
        
        // Count current cages
        int current_cages = 0;
        for(int i=0; i<TOTAL_CELLS; ++i) if(uf.parent[i] == i) current_cages++;
        if (current_cages <= target_count) break;

        cout << "Round " << round << ": " << current_cages << " cages. Target: " << target_count << endl;

        bool progress = false;
        
        // Slowly increase allowed merge size to force small cages first
        int max_merge_size = 2 + (round / 2); 
        if (max_merge_size > info.max_cage) max_merge_size = info.max_cage;
        
        int merges_tried = 0;
        int merges_success = 0;
        
        for (auto& edge : edges) {
            if (current_cages <= target_count) break;
            
            int u = edge.first;
            int v = edge.second;
            int root_u = uf.find(u);
            int root_v = uf.find(v);
            
            if (root_u == root_v) continue;
            if (uf.size[root_u] + uf.size[root_v] > max_merge_size) continue;

            // Tentative merge
            int old_parent_v = uf.parent[root_v];
            int old_size_u = uf.size[root_u];
            
            uf.parent[root_v] = root_u;
            uf.size[root_u] += uf.size[root_v];
            
            merges_tried++;

            // Optimization: if merging two singletons (1+1=2), it's almost always unique.
            // Skip check for 1+1 merges to speed up early phase.
            // Actually, let's skip check if resulting size <= 2.
            if (uf.size[root_u] <= 2) {
                 current_cages--;
                 progress = true;
                 merges_success++;
                 continue;
            }

            // Also skip check if resulting size <= 3 and we are in early rounds (0-5)
            // This is a heuristic: size 3 cages are also very restrictive.
            if (uf.size[root_u] <= 3 && round < 5) {
                 current_cages--;
                 progress = true;
                 merges_success++;
                 continue;
            }
            
            // Skip check for size 4 in very early rounds (0-2)
            if (uf.size[root_u] <= 4 && round < 2) {
                 current_cages--;
                 progress = true;
                 merges_success++;
                 continue;
            }
            
            // Even more aggressive: if size <= 3, skip check always.
            // Size 3 cages (e.g. sum 6 -> 1,2,3) are very strong.
            // The risk is low.
            if (uf.size[root_u] <= 3) {
                 current_cages--;
                 progress = true;
                 merges_success++;
                 continue;
            }

            vector<Cage> test_cages = cages_from_uf(uf, board);
            if (solver.has_unique_solution(test_cages)) {
                current_cages--;
                progress = true;
                merges_success++;
            } else {
                // Undo
                uf.parent[root_v] = old_parent_v;
                uf.size[root_u] = old_size_u;
            }
        }
        cout << "  Tried: " << merges_tried << ", Success: " << merges_success << endl;
        if (!progress && max_merge_size >= info.max_cage) break;
    }

    // Phase 2: Force merge undersized cages
    bool changed = true;
    while (changed) {
        changed = false;
        for (int i = 0; i < TOTAL_CELLS; i++) {
            int root = uf.find(i);
            if (uf.parent[i] == i && uf.size[root] < info.min_cage) {
                // Find neighbor to merge
                vector<int> neighbors;
                // Scan all cells in this cage to find neighbors
                for(int k=0; k<TOTAL_CELLS; ++k) {
                    if(uf.find(k) == root) {
                        int r = k/9, c = k%9;
                        int nbs[4][2] = {{r+1,c}, {r-1,c}, {r,c+1}, {r,c-1}};
                        for(auto& nb : nbs) {
                            if(nb[0]>=0 && nb[0]<9 && nb[1]>=0 && nb[1]<9) {
                                int nb_idx = nb[0]*9 + nb[1];
                                int nb_root = uf.find(nb_idx);
                                if(nb_root != root) {
                                    neighbors.push_back(nb_idx);
                                }
                            }
                        }
                    }
                }
                
                shuffle(neighbors.begin(), neighbors.end(), rng);
                for(int nb_idx : neighbors) {
                    int nb_root = uf.find(nb_idx);
                    if (uf.size[root] + uf.size[nb_root] > info.max_cage + 1) continue; // Allow slightly larger if forcing
                    
                    int old_parent = uf.parent[nb_root];
                    int old_size = uf.size[root];
                    
                    uf.parent[nb_root] = root;
                    uf.size[root] += uf.size[nb_root];
                    
                    if (solver.has_unique_solution(cages_from_uf(uf, board))) {
                        changed = true;
                        goto next_cage;
                    } else {
                        uf.parent[nb_root] = old_parent;
                        uf.size[root] = old_size;
                    }
                }
            }
            next_cage:;
        }
    }

    // Final validation
    for (int i = 0; i < TOTAL_CELLS; i++) {
        if (uf.parent[i] == i) {
            if (uf.size[i] < info.min_cage || uf.size[i] > info.max_cage + 1) return {};
        }
    }

    return cages_from_uf(uf, board);
}

// ============================================================================
// Main & JSON Output
// ============================================================================

string cage_to_string(const Cage& c) {
    string s = "{\"sum\": " + to_string(c.sum) + ", \"cells\": [";
    for (size_t i = 0; i < c.cells.size(); i++) {
        int r = c.cells[i] / 9;
        int col = c.cells[i] % 9;
        s += "[" + to_string(r) + "," + to_string(col) + "]";
        if (i < c.cells.size() - 1) s += ",";
    }
    s += "]}";
    return s;
}

string board_to_string(const vector<vector<int>>& b) {
    string s = "[";
    for (int i = 0; i < 9; i++) {
        s += "[";
        for (int j = 0; j < 9; j++) {
            s += to_string(b[i][j]);
            if (j < 8) s += ",";
        }
        s += "]";
        if (i < 8) s += ",";
    }
    s += "]";
    return s;
}

int main(int argc, char* argv[]) {
    int count_per_level = 1;
    if (argc == 3 && string(argv[1]) == "--count") {
        count_per_level = stoi(argv[2]);
    } else {
        cout << "Usage: ./generate_killer --count N" << endl;
        return 1;
    }

    // Load existing (mock implementation - assumes empty or append)
    // In a real C++ implementation, parsing JSON is heavy. 
    // We will output a partial JSON fragment that can be manually merged or valid JSON if file is empty.
    
    cout << "Generating " << count_per_level << " puzzles per level..." << endl;

    // We'll write to a temp file and let Python handle the JSON merging if needed, 
    // or just write a valid JSON structure here.
    // Let's write a valid JSON structure.
    
    ofstream out(OUTPUT_FILE);
    out << "{\"puzzles\": [\n";

    int global_id = 1;
    bool first = true;

    for (auto const& [level, info] : DIFFICULTY_LEVELS) {
        cout << "\n--- Level " << level << " (" << info.name << ") ---" << endl;
        
        for (int i = 0; i < count_per_level; i++) {
            auto start_time = chrono::high_resolution_clock::now();
            int attempts = 0;
            
            while (true) {
                attempts++;
                if (attempts % 10 == 0) cout << "." << flush;
                
                auto board = generate_full_board();
                auto cages = generate_cages(board, level);
                
                if (!cages.empty()) {
                    if (!first) out << ",\n";
                    first = false;
                    
                    out << "  {\n";
                    out << "    \"id\": " << global_id++ << ",\n";
                    out << "    \"difficulty\": " << level << ",\n";
                    out << "    \"difficulty_name\": \"" << info.name << "\",\n";
                    out << "    \"cages\": [\n";
                    for (size_t k = 0; k < cages.size(); k++) {
                        out << "      " << cage_to_string(cages[k]);
                        if (k < cages.size() - 1) out << ",";
                        out << "\n";
                    }
                    out << "    ],\n";
                    out << "    \"solution\": " << board_to_string(board) << "\n";
                    out << "  }";
                    
                    auto end_time = chrono::high_resolution_clock::now();
                    chrono::duration<double> elapsed = end_time - start_time;
                    
                    cout << "  Generated #" << (global_id - 1) << " (" << cages.size() 
                         << " cages, " << fixed << setprecision(2) << elapsed.count() << "s, " 
                         << attempts << " attempts)" << endl;
                    break;
                }
            }
        }
    }
    out << "\n]}\n";
    out.close();
    cout << "\nDone! Saved to " << OUTPUT_FILE << endl;

    return 0;
}

File: generate_killer.py
import argparse
import json
import os
import random
import sys
import time
from itertools import combinations

sys.setrecursionlimit(50000)

SIZE = 9
TOTAL_CELLS = SIZE * SIZE
DIGITS = frozenset(range(1, SIZE + 1))
CELLS = [(r, c) for r in range(SIZE) for c in range(SIZE)]

DIFFICULTY_LEVELS = {
    1: {"name": "Beginner", "min_cage": 2, "max_cage": 4, "target_avg": 2.7},
    2: {"name": "Easy",     "min_cage": 3, "max_cage": 5, "target_avg": 3.3},
    3: {"name": "Medium",   "min_cage": 3, "max_cage": 5, "target_avg": 3.8},
    4: {"name": "Hard",     "min_cage": 3, "max_cage": 5, "target_avg": 4.3},
    5: {"name": "Expert",   "min_cage": 4, "max_cage": 6, "target_avg": 5.0},
}

OUTPUT_FILE = "sudoku_killer.json"


# ---------------------------------------------------------------------------
# Precompute valid digit combinations: COMBOS[(n, s)] -> list of frozensets
# For cage of size n with target sum s, all subsets of {1..9} that work.
# ---------------------------------------------------------------------------

COMBOS = {}


def _precompute_combos():
    for n in range(1, 10):
        for combo in combinations(range(1, 10), n):
            s = sum(combo)
            COMBOS.setdefault((n, s), []).append(frozenset(combo))


_precompute_combos()

# ---------------------------------------------------------------------------
# Grid adjacency
# ---------------------------------------------------------------------------

_DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def _cell_neighbors(r, c):
    out = []
    for dr, dc in _DIRS:
        nr, nc = r + dr, c + dc
        if 0 <= nr < SIZE and 0 <= nc < SIZE:
            out.append((nr, nc))
    return out


_NBR = {cell: _cell_neighbors(*cell) for cell in CELLS}

# ---------------------------------------------------------------------------
# Standard topology (rows, cols, 3x3 boxes) -- fixed for all puzzles
# ---------------------------------------------------------------------------

_ALL_UNITS = []
for _r in range(SIZE):
    _ALL_UNITS.append(tuple(((_r, _c) for _c in range(SIZE))))
for _c in range(SIZE):
    _ALL_UNITS.append(tuple(((_r, _c) for _r in range(SIZE))))
for _br in range(0, SIZE, 3):
    for _bc in range(0, SIZE, 3):
        _ALL_UNITS.append(
            tuple((_br + dr, _bc + dc) for dr in range(3) for dc in range(3))
        )

_UNITS = {}
_PEERS = {}
for _cell in CELLS:
    _UNITS[_cell] = [u for u in _ALL_UNITS if _cell in u]
    _ps = set()
    for _u in _UNITS[_cell]:
        _ps.update(_u)
    _ps.discard(_cell)
    _PEERS[_cell] = frozenset(_ps)

# ---------------------------------------------------------------------------
# Board generation (backtracking, ported from generate.py)
# ---------------------------------------------------------------------------


def _is_valid(board, row, col, num):
    if num in board[row]:
        return False
    if any(board[r][col] == num for r in range(SIZE)):
        return False
    br, bc = 3 * (row // 3), 3 * (col // 3)
    for r in range(br, br + 3):
        for c in range(bc, bc + 3):
            if board[r][c] == num:
                return False
    return True


def generate_full_board():
    board = [[0] * SIZE for _ in range(SIZE)]

    def fill(pos):
        if pos == TOTAL_CELLS:
            return True
        row, col = divmod(pos, SIZE)
        nums = list(range(1, SIZE + 1))
        random.shuffle(nums)
        for num in nums:
            if _is_valid(board, row, col, num):
                board[row][col] = num
                if fill(pos + 1):
                    return True
                board[row][col] = 0
        return False

    fill(0)
    return board


# ---------------------------------------------------------------------------
# Cage generation (Union-Find random merge)
# ---------------------------------------------------------------------------



def _cell_index(r, c):
    return r * SIZE + c




def generate_cages(solution, difficulty):
    """Build cages by starting from 81 singletons and iteratively merging
    adjacent cages while preserving unique solvability."""
    info = DIFFICULTY_LEVELS[difficulty]
    min_cage = info["min_cage"]
    max_cage = info["max_cage"]
    target_avg = info["target_avg"]
    target_count = max(int(round(TOTAL_CELLS / target_avg)), 9)

    cage_of = list(range(TOTAL_CELLS))
    cage_members = {i: {i} for i in range(TOTAL_CELLS)}

    edges = []
    for r, c in CELLS:
        idx = _cell_index(r, c)
        for nr, nc in _NBR[(r, c)]:
            nidx = _cell_index(nr, nc)
            if idx < nidx:
                edges.append((idx, nidx))

    def _merge(ca, cb):
        if len(cage_members[ca]) < len(cage_members[cb]):
            ca, cb = cb, ca
        for idx in cage_members[cb]:
            cage_of[idx] = ca
        cage_members[ca] |= cage_members[cb]
        old_cb = cage_members.pop(cb)
        return ca, cb, old_cb

    def _undo_merge(ca, cb, old_cb):
        for idx in old_cb:
            cage_of[idx] = cb
        cage_members[ca] -= old_cb
        cage_members[cb] = old_cb

    def _to_cage_list():
        cages = []
        for cid in sorted(cage_members.keys()):
            cells = []
            cage_sum = 0
            for idx in sorted(cage_members[cid]):
                r, c = divmod(idx, SIZE)
                cells.append([r, c])
                cage_sum += solution[r][c]
            cages.append({"cells": cells, "sum": cage_sum})
        return cages

    # Merge iteratively, prioritizing smaller cages
    for _round in range(10):
        # Sort edges by combined size of cages they connect
        edges.sort(key=lambda x: len(cage_members[cage_of[x[0]]]) + len(cage_members[cage_of[x[1]]]) + random.random())
        
        if len(cage_members) <= target_count:
            break
            
        progress = False
        batch = []
        batch_size = 3
        
        for a, b in edges:
            if len(cage_members) <= target_count:
                break
            ca, cb = cage_of[a], cage_of[b]
            if ca == cb:
                continue
            if len(cage_members[ca]) + len(cage_members[cb]) > max_cage:
                continue

            merged_ca, merged_cb, old_cb = _merge(ca, cb)
            batch.append((merged_ca, merged_cb, old_cb))
            
            if len(batch) >= batch_size:
                if has_unique_solution(_to_cage_list()):
                    progress = True
                    batch = []
                else:
                    # Batch failed, undo all
                    for m_ca, m_cb, m_old in reversed(batch):
                        _undo_merge(m_ca, m_cb, m_old)
                    batch = []
        
        # Process remaining batch
        if batch:
            if has_unique_solution(_to_cage_list()):
                progress = True
            else:
                for m_ca, m_cb, m_old in reversed(batch):
                    _undo_merge(m_ca, m_cb, m_old)
        
        # print(f"Round {_round}: {len(cage_members)} cages", flush=True)
        if not progress:
            break

    # Phase 2: force-merge undersized cages
    changed = True
    while changed:
        changed = False
        for cid in list(cage_members.keys()):
            if cid not in cage_members:
                continue
            if len(cage_members[cid]) >= min_cage:
                continue
            merged = False
            for member_idx in list(cage_members[cid]):
                mr, mc = divmod(member_idx, SIZE)
                for nr, nc in _NBR[(mr, mc)]:
                    nidx = _cell_index(nr, nc)
                    ncid = cage_of[nidx]
                    if ncid == cid:
                        continue
                    if len(cage_members[cid]) + len(cage_members[ncid]) \
                            > max_cage:
                        continue
                    m_ca, m_cb, m_old = _merge(cid, ncid)
                    if has_unique_solution(_to_cage_list()):
                        changed = True
                        merged = True
                        break
                    else:
                        _undo_merge(m_ca, m_cb, m_old)
                if merged:
                    break

    # Validate: all cages must meet size constraints
    for cid, members in cage_members.items():
        if len(members) < min_cage or len(members) > max_cage:
            return None
    return _to_cage_list()


# ---------------------------------------------------------------------------
# Killer Sudoku solver (constraint propagation + backtracking)
# ---------------------------------------------------------------------------



def _build_cage_index(cages):
    """Map each cell to its cage index, and build cage data structures."""
    cell_to_cage = {}
    cage_cells = []
    cage_sums = []
    for ci, cage in enumerate(cages):
        cells = [tuple(c) for c in cage["cells"]]
        cage_cells.append(cells)
        cage_sums.append(cage["sum"])
        for cell in cells:
            cell_to_cage[cell] = ci
    return cell_to_cage, cage_cells, cage_sums



_SOLVER_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                            "killer_solver")

import subprocess


class _SolverProcess:
    """Persistent C solver process for fast uniqueness checks."""

    def __init__(self):
        self._proc = None

    def _ensure(self):
        if self._proc is None or self._proc.poll() is not None:
            self._proc = subprocess.Popen(
                [_SOLVER_PATH],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                text=True,
                bufsize=1,
            )

    def check(self, cages):
        self._ensure()
        lines = [str(len(cages))]
        for cage in cages:
            cells = cage["cells"]
            cell_indices = " ".join(str(c[0] * SIZE + c[1]) for c in cells)
            lines.append(f"{len(cells)} {cage['sum']} {cell_indices}")
        self._proc.stdin.write("\n".join(lines) + "\n")
        self._proc.stdin.flush()
        result = self._proc.stdout.readline().strip()
        return result == "1"

    def close(self):
        if self._proc and self._proc.poll() is None:
            self._proc.stdin.close()
            self._proc.wait()


_solver = _SolverProcess()


def has_unique_solution(cages):
    """Check uniqueness via persistent compiled C solver."""
    return _solver.check(cages)


# ---------------------------------------------------------------------------
# Persistence
# ---------------------------------------------------------------------------


def cages_to_key(cages):
    parts = []
    for cage in cages:
        cells_tuple = tuple(tuple(c) for c in sorted(cage["cells"]))
        parts.append((cells_tuple, cage["sum"]))
    return tuple(sorted(parts))


def load_existing(filepath):
    if not os.path.exists(filepath):
        return {"puzzles": []}
    with open(filepath, "r", encoding="utf-8") as f:
        return json.load(f)


def save_puzzles(data, filepath):
    with open(filepath, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)


def get_existing_keys(data):
    keys = set()
    for entry in data["puzzles"]:
        keys.add(cages_to_key(entry["cages"]))
    return keys


# ---------------------------------------------------------------------------
# Main generation loop
# ---------------------------------------------------------------------------


def generate_puzzles(count_per_level):
    data = load_existing(OUTPUT_FILE)
    existing_keys = get_existing_keys(data)
    next_id = max((p["id"] for p in data["puzzles"]), default=0) + 1
    total = count_per_level * len(DIFFICULTY_LEVELS)
    generated = 0

    for level in sorted(DIFFICULTY_LEVELS.keys()):
        info = DIFFICULTY_LEVELS[level]
        print(f"\n--- Level {level} ({info['name']}) ---")

        for _ in range(count_per_level):
            t0 = time.time()
            attempts = 0
            while True:
                attempts += 1
                if attempts % 5 == 0:
                    print(f"  (attempt {attempts}...)", flush=True)

                solution = generate_full_board()
                cages = generate_cages(solution, level)
                if cages is None:
                    continue

                key = cages_to_key(cages)
                if key not in existing_keys:
                    existing_keys.add(key)
                    break

            entry = {
                "id": next_id,
                "difficulty": level,
                "difficulty_name": info["name"],
                "cages": cages,
                "solution": solution,
            }
            data["puzzles"].append(entry)
            next_id += 1
            generated += 1
            elapsed = time.time() - t0
            n_cages = len(cages)
            avg_sz = TOTAL_CELLS / n_cages
            print(f"  [{generated}/{total}] #{entry['id']} "
                  f"({n_cages} cages, avg {avg_sz:.1f}, "
                  f"{elapsed:.1f}s, {attempts} attempts)")

    save_puzzles(data, OUTPUT_FILE)
    print(f"\nDone! {generated} puzzles saved to {OUTPUT_FILE}")
    print(f"Total puzzles in file: {len(data['puzzles'])}")


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def main():
    parser = argparse.ArgumentParser(
        description="Generate Killer Sudoku puzzles")
    parser.add_argument(
        "--count", type=int, required=True,
        help="Number of puzzles to generate per difficulty level",
    )
    args = parser.parse_args()
    if args.count < 1:
        print("Error: --count must be at least 1")
        sys.exit(1)

    print(
        f"Generating {args.count} puzzle(s) per level x "
        f"{len(DIFFICULTY_LEVELS)} levels "
        f"= {args.count * len(DIFFICULTY_LEVELS)} total"
    )
    start = time.time()
    generate_puzzles(args.count)
    print(f"Total time: {time.time() - start:.1f}s")


if __name__ == "__main__":
    main()
